File: plugin/octo_beat_plugin/ios/Classes/event/EventHandler.swift
//
//  EventHandler.swift
//  octo_beat_plugin
//
//  Created by TRAN QUANG DAI on 09/02/2023.
//

import Foundation
import Flutter

class EventHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    func send(event: String, body: [String: Any]?) {
        eventSink?(["event": event, "body": body])
    }
    
    func send(event: String, body: [AnyHashable: Any]) {
        eventSink?(["event": event, "body": body])
    }
}

File: plugin/octo_beat_plugin/android/src/main/kotlin/com/octo/octo_beat_plugin/plugin/event/EventHandler.kt
package com.octo.octo_beat_plugin.plugin.event

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.octo.octo_beat_plugin.plugin.service.OctoTask
import io.flutter.plugin.common.EventChannel

class EventHandler : EventChannel.StreamHandler {
    private var eventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    fun send(task: OctoTask) {
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(mapOf(
                "event" to task.method,
                "body" to task.data
            ))

        }
    }

    fun send(event: String, body: Map<String, Any>?) {
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(mapOf("event" to event, "body" to body))
        }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}